### PR TITLE
Add Markdoc preset and example

### PR DESCRIPTION
**Pull request by HiDeoo**
#### Description

This PR adds a Markdoc preset and example to Starlight so that users can more easily use or get started with Markdoc in Starlight.

#### Limitations

There are a few limitations in the current version due to either Markdoc itself or other constraints. They are all documented in the preset itself but I will list the most important ones here:

- `icon` attributes do not have an explicit set of `matches` and just accept any string.
- A few components that accept extra props, e.g. all attributes of the `<a>` tag, ~~don't have the same parity in Markdoc~~.
	- The associated tags now support a list of well-known attributes that we can improve over time.
- Fenced code blocks do not support markers as Markdoc ignore them, so user would need to use the `code` tag instead with the `meta` attribute.

I'll try to see if there are some valid workarounds for these limitations.

~~There is also an issue that we cannot use the `extends` field in a `markdoc.config.mjs` file. Using it lead to an `Unable to render [object Object]. No valid renderer was found for this file extension.` error.~~
~~I'll have to investigate this issue further to see if it's a bug or a real limitation, e.g. maybe not being able to use `component` in the `extends` field for some reason.~~
Opened https://github.com/withastro/astro/pull/11846 to fix this issue.

#### Testing

We briefly discussed about ensuring some kind of sync between Starlight user components and their tags equivalent in Markdoc. I opted to use type testing as I think it may be one of the easiest approach to this.

We can rely on the amazing [`ComponentProps`](https://docs.astro.build/en/guides/typescript/#componentprops-type) type from Astro to extract user components props and TypeScript type-inference on the preset (that only use `satisfies` to keep the most specific type) to extract Markdoc tags attributes. We can then compare the two types to ensure they are in sync.

At the moment, there are 2 type of type tests:

- We test that all user components have an equivalent Markdoc tag in the Markdoc preset.
- We test that each Markdoc tag have attributes that matches the props of the equivalent user component (we don't test the type of the attributes, only their presence so far. Not sure yet if we should test the type as well as this may add a lot more complexity).

We rely on a few Vitest type testing utilities but we do not run these tests using Vitest as type tests seems a bit fragile when using Vitest workspaces, and also because Vitest would use `tsc` to run those and this would not play well with Astro components. Instead, the already existing `pnpm typecheck` command will ensure these tests are valid.

#### Documentation

~~Need to think a bit more about this. I think we should at least have documentation on how to use the new Markdoc preset. Not sure yet tho where this documentation should be placed, the "Authoring Content in Markdown" page may not be the best place for this.~~

I added a new Markdoc documentation section using the same structure used for `@astrojs/starlight-tailwind`.

- Currently at the bottom of the "Authoring Content in Markdown" guide (not quite sure yet if it's the best place for it)
- I purposefully kept some unrelated formatting changes in the document as it was changed from `.md` to `.mdx` and we know how Prettier & MDX can be a bit tricky sometimes so I wanted to make sure nothing would break when the formatter would run.

#### Remaining tasks

- [x] Wait for https://github.com/withastro/astro/pull/11846 to be merged and released
	- [x] Update `packages/markdoc/package.json` file `peerDependencies` to use the new `@astrojs/markdoc` version
	- [x] Update `examples/markdoc/markdoc.config.mjs` to use the `extends` syntax
- [x] Documentation
	- [x] Figure out the best place for it
	- [x] When done with the documentation, update the link to it in `packages/markdoc/README.md`
- [x] Do we want a GitHub label for this new package
	- [x] If yes, `.github/labeler.yml` should be updated
- [x] Remove sidebar demo group from `examples/markdoc/astro.config.mjs`
- [x] Remove the `examples/markdoc/src/content/docs/demo/` folder and all its content
- [x] Add a changeset

### CODE REVIEW

**Review by delucis — COMMENTED** [review]
Thanks for working on this @HiDeoo!

Quick meta question ahead of a more thorough review: do you think this might make more sense as a standalone package Markdoc users need to install? As they will in any case need to add some manual config, the extra friction to `npm i @astrojs/starlight-markdoc-preset` or whatever the package name is, doesn’t seem too bad and would help keep things separate.

**Comment by HiDeoo** [comment]
> Quick meta question ahead of a more thorough review: do you think this might make more sense as a standalone package Markdoc users need to install? As they will in any case need to add some manual config, the extra friction to `npm i @astrojs/starlight-markdoc-preset` or whatever the package name is, doesn’t seem too bad and would help keep things separate.

Good question, I remember when we quickly discussed this we mentioned the possibility of having a dedicated package for this.

The current implementation lives in the core package as it was easier to setup everything but I made sure that everything pretty much depends only on `@astrojs/starlight/components`, even the tests. So it should be easy to move it to a dedicated package if we want to.

- For a new user using the Markdoc example, this would not change anything, just an extra dependency installed automatically.
- For an existing user adding Markdoc manually, they would install `@astrojs/markdoc` already so adding the extra dependency would not change a lot.
- For an existing user adding Markdoc using `astro add`, they would need an extra different command to install the package.

I'm personally leaning towards having a dedicated package for this, as I like the separation (just like `@astrojs/markdoc` being a separate package), but I don't have a super strong opinion on this. I'm happy to do the work if we decide to go this route, should be pretty straightforward.

**Comment by HiDeoo** [comment]
Updated the PR with the following changes:

- Move everything to a dedicated `@astrojs/starlight-markdoc` package
- Add support for well-known HTML attributes to various tags
- Add a new Markdoc documentation section using the same structure used for `@astrojs/starlight-tailwind`
  - Currently at the bottom of the "Authoring Content in Markdown" guide (not quite sure yet if it's the best place for it)
  - I purposefully kept some unrelated formatting changes in the document as it was changed from `.md` to `.mdx` and we know how Prettier & MDX can be a bit tricky sometimes so I wanted to make sure nothing would break when the formatter would run on `main`.

I also opened a [PR](https://github.com/withastro/astro/pull/11846) on the `astro` repo to fix the `extends` issue I mentioned. We would need to wait for that to be merged and released before releasing the new preset (and make sure the `peerDependencies` version is updated accordingly).

**Review by delucis — APPROVED** [review]
LGTM @HiDeoo! Left a couple of tiny details but I don’t have any major feedback. I think once we have #2121 merged, it would be nice to have a MDX / Markdoc tab to show the component syntax probably?

But that’s for the future.

File: examples/markdoc/.vscode/extensions.json
Comment (HiDeoo) [reply examples/markdoc/.vscode/extensions.json:2]: Thought about it but I then wondered if we should also include the [`markdoc.config.json` file](https://docs.astro.build/en/guides/integrations-guide/markdoc/#vs-code-editor-integration) required by the extension, but if we do, users would get validation errors for all tags due to the extension not supporting `extends` and this made me doubt if we should include it or not.

Or maybe just the recommendation is fine without the config?

**Comment by HiDeoo** [comment]
> I think once we have #2121 merged, it would be nice to have a MDX / Markdoc tab to show the component syntax probably?

Definitely, as soon as this PR is released, I will update #2121 to include the MDX and Markdoc syntax. It would still use the same component but with the Markdoc syntax in a new different slot (altho, the preview will still be the MDX one as this wouldn't change anything). I will also need to research a bit about Markdoc common formatting, if they tend to indent nested tags or not, etc. as I'm not familiar with it but we should provide examples in a format that is common to the users.

Thanks for the review :raised_hands: I'll tackle the remaining points early next week.

File: examples/markdoc/.vscode/extensions.json
Comment (delucis) [reply examples/markdoc/.vscode/extensions.json:2]: Doesn’t the extension at least provide syntax highlighting? Or does VS Code have that out of the box? I tried it locally and saw syntax highlighting for tags and attributes so assumed that was the extension.

Comment (HiDeoo) [reply examples/markdoc/.vscode/extensions.json:2]: Oh, you are totally right, just tested it and the extension provides syntax highlighting even without its configuration file. I guess it's only used for the LSP and validation/completions. I wrongly assumed it was mandatory for everything.

In that case, I agree that we should include it in the list 👍

**Review by delucis — APPROVED** [review]
Lovely work! Ready for launch 🚀

**Comment by delucis** [comment]
Updated the Markdoc example to the latest Astro to match the changes in #2287 and still LGTM!

**Pull request merged**
